common/src/main/java/feast/common/interceptors/GrpcMessageInterceptor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2019 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.common.interceptors;
+
+import com.google.protobuf.Message;
+import feast.common.logging.AuditLogger;
+import feast.common.logging.LogEvent;
+import feast.common.logging.LogEvent.MessageLogEvent.Direction;
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
+import io.grpc.ForwardingServerCallListener.SimpleForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import org.apache.logging.log4j.Level;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+/**
+ * GrpcMessageInterceptor intercepts a GRPC calls to log handling of GRPC messages to the Audit Log.
+ * Intercepts the incoming and outgoing messages logs them to the audit log, together with method
+ * name and assumed authenticated identity (if authentication is enabled).
+ */
+public class GrpcMessageInterceptor implements ServerInterceptor {
+  private AuditLogger auditLog;
+
+  public GrpcMessageInterceptor(AuditLogger auditLog) {
+    this.auditLog = auditLog;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    // Unpack method name from call
+    String fullMethodName = call.getMethodDescriptor().getFullMethodName();
+    String methodName = fullMethodName.substring(fullMethodName.indexOf("/") + 1);
+
+    // Attempt Extract current authenticated identity.
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+    String identity = (authentication == null) ? "" : authentication.getPrincipal().toString();
+
+    // Register forwarding call to intercept outgoing response and log to audit log
+    call =
+        new SimpleForwardingServerCall<ReqT, RespT>(call) {
+          @Override
+          public void sendMessage(RespT message) {
```

Comment:
Is this supposed to apply to GetOnlineFeatures as well, because in that case we will probably experience a serious performance hit. Audit logging is normally distinct from request/response logging, since with the former you expect durability and consistency but with the latter you want availability and performance.

Reply:
The `AuditLogger` can be configured via `application.yml` via `AuditLogProperties` and thus can be disabled for the Online Serving instance from `application.yml` for cases where latency is more important that visibility.

Reply:
Yep, but we need request/response logging in online serving as well. We just want to make sure that this logging happens in async fashion and doesn't affect latency of the method